File: 백준/Gold/13023. ABCDE/ABCDE.cc
#include <iostream>
#include <vector>

using namespace std;

int N, M; 
vector<vector<int>> Graph;
vector<bool> Visited;

bool DFS(int Src, int Depth = 0)
{
    if(Depth == 4)
    {
        return true;
    }
    
    Visited[Src] = true;
    for(int Dst : Graph[Src])
    {
        if(Visited[Dst])
        {
            continue; 
        }
        
        if(DFS(Dst, Depth + 1))
        {
            return true;
        }
    }
    Visited[Src] = false;
    
    return false;
}

int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);
    
    cin >> N >> M;
    Graph.assign(N, {});
    Visited.assign(N, false);
    
    while(M--)
    {
        int A, B; cin >> A >> B;
        Graph[A].emplace_back(B);
        Graph[B].emplace_back(A);
    }
    
    bool Result = false;
    for(int Node = 0; Node < N; ++Node)
    {
        if(Visited[Node])
        {
            continue;
        }
        
        Result = DFS(Node);
        if(Result)
        {
            break;
        }
    }
    
    cout << Result;
    
    return 0;
}

File: 백준/Gold/13023. ABCDE/ABCDE.py
input = open(0).readline

N, M = map(int, input().split())
graph = [[] for _ in range(N)]

for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    
visited = [False] * N


def dfs(src: int, depth: int = 0) -> bool:
    if depth == 4:
        return True
        
    visited[src] = True
    for dst in graph[src]:
        if visited[dst]:
            continue
        
        if dfs(dst, depth + 1):
            return True
            
    visited[src] = False
    return False
    

def solution() -> bool:
    for node in range(N):
        if visited[node]:
            continue
        
        if dfs(node):
            return True
            
    return False


print('1' if solution() else '0')